verilog/analysis/checkers/numeric_format_string_style_rule.cc

```diff
@@ -0,0 +1,143 @@
+#include "verilog/parser/verilog_token_enum.h"
+
+namespace verilog {
+namespace analysis {
+
+using verible::GetStyleGuideCitation;
+using verible::LintRuleStatus;
+using verible::LintViolation;
+using verible::TokenInfo;
+using verible::TokenStreamLintRule;
+
+// Register the lint rule
+VERILOG_REGISTER_LINT_RULE(NumericFormatStringStyleRule);
+
+absl::string_view NumericFormatStringStyleRule::Name() { return "numeric-format-string-style"; }
+const char NumericFormatStringStyleRule::kTopic[] = "format-string";
+const char NumericFormatStringStyleRule::kMessage[] =
+  "Formatting string must contain proper style-compilant numeric specifiers.";
+
+std::string NumericFormatStringStyleRule::GetDescription(
+    DescriptionType description_type) {
+  return absl::StrCat("Checks that string literals with numeric format specifiers "
+    "have proper prefixes for hex and bin values and no prefixes for decimal values.");
+}
+
+void NumericFormatStringStyleRule::HandleToken(const TokenInfo& token) {
+  const auto token_enum = static_cast<verilog_tokentype>(token.token_enum());
+  const absl::string_view text(token.text());
+
+  if (token_enum != TK_StringLiteral)
+    return ;
+
+  for (size_t pos = 0 ; pos < text.size() ;) {
+    const auto& ch = text[pos];
+
+    // Skip ordinary characters
+    if (ch != '%') {
+      ++pos;
+      continue ;
+    }
+
+    // Examine formatting directive
+    for (size_t itr = pos + 1 ; itr < text.size() ; ++itr) {
+      // Skip field with display data size
+      if (absl::ascii_isdigit(text[itr]))
+        continue ;
+
+      bool violation = false;
+
+      const auto radix = text[itr];
+
+      // Format radix
+      switch (radix) {
+        // binary value
+        case 'b':
+        case 'B':
+
+        // hexdecimal value
+        case 'x':
+        case 'X': {
+          // Guilty until proven innocent
+          violation = true;
+
+          // Check for prefix
+          if (pos >= 2 &&
+              text[pos-1] == absl::ascii_tolower(radix) &&
+              (text[pos-2] == '0' || text[pos-2] == '\'')) {
+            violation = false;
+          }
+          break;
+        }
+
+        // decimal value
+        case 'd':
+        case 'D': {
+          // Innocent until proven guilty
+          violation = false;
+
+          // TODO(fangism): Detecting only 4 prefixes.
+          //   Should it be more generic?, e.g.
+          //   if (pos >= 2 && absl::ascii_isalpha(text[pos-1]))
+          //     violation = true
+          if (pos >= 2 &&
+              text[pos-1] == absl::ascii_tolower(radix) &&
+              (text[pos-2] == '0' || text[pos-2] == '\'')) {
+            violation = true;
+          }
+          break;
+        }
+
+        default:
+          break;
+      }
+
+      // TODO(fangism): Different message for different formatting string violations?
+      if (violation)
+        violations_.insert(LintViolation(token, kMessage));
```

Comment:
Instead of the whole token, can we synthesize a token whose .text() spans the offending format specifier (and its prefix where applicable).  `string_view::substr()` should do it.  This would give a very precise location/byte-offset.

Reply:
... and support finding multiple violations in unique locations within one string literal.  :)